File: ruby/lib/oropy_bridge/oropy_bridge.rb
require 'msgpack'
require 'orocos'
include Orocos

# Provides a class that has methods to
#  - start deployments
#  - configure a task
#  - start a task
#  - feed data to a /base/VectorXd Port
#  - read data from a /base/VectorXd Port
#  - connect ports
#  - stop a task
#  - clean up a task
#  - stop deployments
module OropyBridge

    def self.type_to_ruby(value)
        if value.kind_of?(Types::Base::Time)
            value.to_f
        elsif value.kind_of?(Typelib::CompoundType)
            result = Hash.new
            value.each_field do |field_name, field_value|
                result[field_name] = type_to_ruby(field_value)
            end
            result
        elsif value.kind_of?(Symbol)
            value.to_s
        elsif value.respond_to?(:to_str)
            value.to_str
        elsif value.kind_of?(Typelib::ArrayType) || value.kind_of?(Typelib::ContainerType)
            value.raw_each.map(&method(:type_to_ruby))
        elsif value == nil
            nil
        elsif value.respond_to?(:to_a)
            value.to_a.each do |v|
                type_to_ruby(v)
            end
        elsif value.kind_of?(Typelib::Type)
            Typelib.to_ruby(value)
        else
            value
        end
    end

    # Provides the methods to run and feed orocos tasks
    class OrocosRb

        attr_reader :tasks

        # Runs a deplyoment very much like:
        #
        #   Orocos.run deployments do
        #    ...
        #   end
        #
        # It does nothing in the block but waits for a stop condition.
        #
        def initialize
            if !Orocos.initialized?
                Orocos.initialize
            end

            @tasks = Hash.new # store name => task mapping
            @deploy_thread = nil
            @writers = Hash.new # stores writers
            @readers = Hash.new # stores readers
        end

        def set_log_all; @log_all = true; end

        def set_no_log; @log_all = false; end

        # deploys the deployments set with initialize
        def deploy(*deployments)

            @deployments = deployments

            @stop_deployments = false

            running = false

            @deploy_thread = Thread.new do
                Orocos.run *@deployments do
                    if @log_all
                        Orocos.log_all
                    end
                    running = true
                    while !@stop_deployments
                        sleep 0.1
                    end
                end
            end

            # to be sure to find all tasks
            while !running; end

            Orocos.name_service.each_task do |task|
                @tasks[task.name] = task
            end

        end

        # stops the deployments
        def stop_deployments
            @stop_deployments = true
            @deploy_thread.join
            @tasks = Hash.new
            @deploy_thread = nil
            @writers = Hash.new
            @readers = Hash.new
        end

        #  set properties of a task given by the hash called config_hash
        def apply_config(task, config_hash)
            config = Orocos::TaskConfigurations.new(@tasks[task].model)
            config.add("default", config_hash)
            config.apply(@tasks[task], "default")
            nil
        end

        # same as task.configure in ruby run scripts
        def configure(task)
            @tasks[task].configure
        end

        # same as task.start in ruby run scripts
        def start(task)
            @tasks[task].start
        end

        # call an operation
        def operation(task, operation_name, arguments)
            @tasks[task].operation(operation_name).callop(*arguments)
        end

        # Does it like
        #
        #   writer = task.port.writer
        #   writer.write(data)
        #
        # data must be a hash matching the type structure
        def write(task, portname, data)
            if !@writers[[task,portname]]
                port = @tasks[task].port(portname)
                type = port.type_name
                @writers[[task,portname]] = [port.writer, Orocos.registry.get(type)]
            end
            data = Typelib.from_ruby(data,@writers[[task,portname]][1])
            @writers[[task,portname]][0].write data
        end

        # Meant to write to ports with type base::VectorXd.
        # Data has to be an array of numerical values.
        def write_vector(task, portname,data)
            write(task, portname, {'data' => data})
        end

        # Does it like
        #
        #   reader = task.port.reader
        #   data = reader.read
        #
        def read(task, portname, new_data=true)
            if !@readers[[task,portname]]
                port = @tasks[task].port(portname)
                type = port.type_name
                @readers[[task,portname]] = [port.reader, Orocos.registry.get(type)]
            end
            if new_data
                data = @readers[[task,portname]][0].read_new
            else
                data = @readers[[task,portname]][0].read
            end
            OropyBridge.type_to_ruby(data)
        end

        # Directly returns the content of the 'data' field.
        # That could be used in conjunction with base::VectorXd.
        def read_vector(task, portname, new_data=false)
            if data = read(task,portname,new_data)
                return data["data"]
            end
        end

        # connect ports
        def connect(from_task, from_port, to_task, to_port)
            @tasks[from_task].port(from_port).connect_to @tasks[from_task].port(to_port)
        end

        # same as task.stop in ruby run scripts
        def stop(task)
            @tasks[task].stop
        end

        # same as task.cleanup in ruby run scripts
        def cleanup(task)
            @tasks[task].cleanup
        end

        # waits for x seconds
        def sleep(seconds)
            Kernel.sleep(seconds)
        end

    end

    # calls methods of a class requested from an io using msgpack messages
    class Caller

        attr_reader :writer
        attr_reader :unpacker
        attr_reader :handler
        attr_reader :verbose
        attr_reader :batch_mode

        def initialize(read_io, write_io, handler, verbose=false, batch_mode=false)
            @writer = write_io
            @unpacker = MessagePack::Unpacker.new(read_io)
            @handler = handler
            @verbose = verbose
            @batch_mode = batch_mode
            @cmd_list = []
        end

        # process messages
        def process
            begin
                if batch_mode
                    process_batch
                else
                    process_direct
                end
            rescue EOFError => e
                false
            end
        end


        # process messages incomming and execute the commands directly
        # returns false if EOF file reached - the connection was closed
        def process_direct
            unpacker.each do |cmd|
                execute_cmd cmd
            end
            true
        end

        # collect the command and reply
        def process_batch
            unpacker.each do |cmd|
                if cmd[0] == "run"
                    @cmd_list.each { |c| execute_cmd c }
                    @cmd_list.clear
                else
                    @cmd_list << cmd
                end
                msg = cmd.to_msgpack
                writer.write(msg).flush
            end
        end
        
        # execute a command and return the answer
        def execute_cmd(cmd)
            begin
                STDERR.puts "ruby side received cmd #{cmd}" if verbose
                reply = @handler.method(cmd[0]).call(*cmd[1])
                msg = [cmd[0],reply].to_msgpack
                STDERR.puts "ruby replies #{reply}" if verbose
            rescue NameError => e
                msg = ["NameError",[e.inspect,e.to_s,e.backtrace]].to_msgpack
            rescue ArgumentError => e
                msg = ["ArgumentError",[e.inspect,e.to_s,e.backtrace]].to_msgpack
            rescue Exception => e
                msg = ["OtherError",[e.inspect,e.to_s,e.backtrace]].to_msgpack
            end
            writer.write(msg)
            writer.flush
        end
    end

end

File: python/oropy_bridge/oropy_bridge.py
try:
    import msgpack
except ImportError, e:
    print >> stderr, "\x1b[31;1mInstall msgpack for python manually:"
    print >> stderr, "   [apt-get install python-pip]"
    print >> stderr, "    pip install --user msgpack-python\x1b[0m"
    raise e
import subprocess
from sys import stderr

class OrocosRb(object):
    ''' Provides all the methods to interact with orogen tasks.'''

    def __init__(self, cmd_forwarder):
        '''Initialization.

        Args:
            cmd_forwarder: A class with a *process* method that takes a msg and
                forwards it to the serving entity and returns the reply.
        '''
        self.cmder = cmd_forwarder

    def deploy(self, deployments):
        '''Deploy tasks.

        Args:
            deployments: The deployments that contain the tasks. Takes deployment names
                , default deplyoments and options.

        .deploy(["deplyomentA", "deploymentB", { "default::Task" : "taskC", "valgrind" : True }]

        Returns:
            The list of tasks started with the deployments.
        '''
        msg = ["deploy",[deployments]]
        return self.cmder.process(msg)

    def stop_deployments(self):
        '''Stop all deployments. Also stops the ruby deployment thread.'''
        msg = ["stop_deployments",[]]
        return self.cmder.process(msg)

    def apply_config(self, task, config):
        '''Applies a configuration to a task. 

        Args:
            task: The task to apply the configuration to.
            config: A dictionary with the properies of the task as keys.
                Only properties meant to be set need to be inside the dictionary.
        '''
        msg = ["apply_config",[task,config]]
        return self.cmder.process(msg)

    def configure(self, task):
        '''Configure a *task* (like task.configure).'''
        msg = ["configure",[task]]
        return self.cmder.process(msg)

    def start(self, task): 
        '''Start a *task* (like task.start).'''
        msg = ["start",[task]]
        return self.cmder.process(msg)

    def operation(self, task, operation, *args):
        '''Call an operation of a task.

        Args:
            task: Name of the task which has the operation.
            operation: Name of the operation.
            *args: Argument list.
        
        Returns:
            The return value of the operation.
        '''
        msg = ["operation",[task,operation,args]]
        return self.cmder.process(msg)

    def write(self, task, portname, data):
        '''Write data to a port.

        Args:
            task: Name of the task that has the port.
            portname: Name of the port.
            data: Data to write to the port. A dictionary that matches the 
                structure of the porttpye.
        '''
        msg = ["write",[task,portname,data]]
        return self.cmder.process(msg)
    
    def write_vector(self, task, portname, data):
        '''Writes a vector to a port of type base::VectorXd.'''
        msg = ["write_vector", [task, portname, data]]
        return self.cmder.process(msg)
    
    def read(self, task, portname, new_data=False):
        '''Read data from a port.
 
        Args:
            task: Name of the task that has the port.
            portname: Name of the port.
            new_data: If True read only new data.

        Returns:
            The data from the port or None if there were no data.
        '''
        msg = ["read", [task,portname, new_data]]
        return self.cmder.process(msg)

    def read_vector(self, task, portname, new_data=False):
        msg = ["read_vector", [task, portname, new_data]]
        return self.cmder.process(msg)

    def connect(self, from_task, from_port, to_task, to_port):
        '''Connect two ports (not tested). 
            
            like from_task.from_port.connect_to to_task.to_port
        '''
        msg = ["connect",[from_task, from_port, to_task, to_port]]
        return self.cmder.process(msg)

    def stop(self,task):
        '''Stops a task (like task.stop).'''
        msg = ["stop",[task]]
        return self.cmder.process(msg)

    def cleanup(self, task):
        '''Cleans up a task (like task.cleanup).'''
        msg = ["cleanup",[task]]
        return self.cmder.process(msg)

    def sleep(self, seconds):
        '''Sleep for seconds.'''
        msg =["sleep",[seconds]]
        return self.cmder.process(msg)


class Forwarder(object):
    '''Forward the command message from the class OrocosRb to another entity that
    executes the commands.
    '''

    def __init__(self, reader, writer=None):
        '''Initialization.

        Args:
            reader: The replies are expected on that line.
            writer: The commands are wirten to this stream.
        '''
        self.reader = reader
        if writer:
            self.writer = writer
        else:
            self.writer = reader
        
    def process_incoming(self):
        ''' Process incoming message from the reader. Uses msgpack to unpack
            the messages.

            Returns:
                The reply.
        '''
        unpacker = msgpack.Unpacker()
        while True:
            data = self.reader.read(1)
            if not data:
                break
            unpacker.feed(data)
            for o in unpacker:
                return o
        return None

    def process(self, msg):
        '''Send messages and wait for the reply.

        Args:
            msg: The command message to forward.

        Returns:
            The reply or None if an error occured.
        '''
        self.writer.write(msgpack.packb(msg))
        reply = self.process_incoming()

        if not reply:
            raise Exception("Something went wrong (no message or timeout)!")
        elif reply[0] == msg[0]:
            return reply[1]
        elif reply[0][-5:] == "Error":
            err_data = reply[1]
            print >> stderr, "\x1b[31;1mError:%s, %s\x1b[0m"%(err_data[0],err_data[1])
            for l in err_data[2]:
                print >> stderr, l
            print >> stderr
            return None


class BatchForwarder(Forwarder):
    '''Batch run all commands.

    Command are collected at the ruby site. And once called run() they will be
    executed. 
    '''

    def get_results(self):
        result = []
        while True:
            reply = self.process_incoming()
            if reply:
                if reply[0] == "run":
                    break
                else:
                    result.append(reply)
            else:
                break
        return result

    def run(self):
        msg = ["run"]
        self.writer.write(msgpack.packb(msg))
        return self.get_results()





class Client(OrocosRb):
    ''' The client to connect to the ruby server to access orocor.rb. '''

    def __init__(self, ruby_cmd="oropy_server", batch_mode=False):
        '''Initializing the client.

        Start the ruby orocos serving side, connect to it and accept commands.
        Commands the class OrocosRb.

        Args:
            ruby_cmd: the command to run the ruby server.
            batch_mode: run in batch mode, collect commands and execute at once.
                Adds "--batch" option to the ruby_cmd.

        A ruby script is started in another process. Both process are connected by a
        pipe. The data serialization is done with msgpack.
        '''
        if ruby_cmd is str:
            ruby_cmd = ruby_cmd.split(" ")
        if batch_mode:
            ruby_cmd.append("--batch") 
        slave = subprocess.Popen(ruby_cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=None)
        if batch_mode:
            super(Client, self).__init__(BatchForwarder(slave.stdout, slave.stdin))
        else:
            super(Client, self).__init__(Forwarder(slave.stdout, slave.stdin))


    def run(self):
        '''Run all comannds when in batch mode.

        Returns:
            The result of the run or None if has not been started in batch mode.
        '''
        if "run" in dir(self.cmder):
            return self.cmder.run()
